### PR TITLE
winograd convolution (for portable and fast)

### PR DESCRIPTION
**Pull request by tsyj0404**
Dear developer, 
thanks for your contribution to this excellent program, here are parts of my work about winograd convolution for cpu inference.

Three files are added for cpu inference based on winograd convolution, winograd_inference.hpp,  winograd_layer_inference. hpp, and winograd_layer_inference.cpp.

1. Three winograd algrithms are supported, and WT8X8 with F(6x6, 3x3) are added to support cpu inference. 
2. Also, winograd2D_initialize() is provided, which supports precomputed and stored before runtime inference.  
3. Float calculations are supported for fast inference.

The new wiongrad interface is more portable for usage, and usages can be found in winograd_layer_inference.cpp.

Hope for a contribution.
Thanks.

### CODE REVIEW

**Comment by jspark1105** [comment]
Thanks for the contribution. I think your contribution on dense Winograd convolution can be more appreciated in the main caffe branch because SkimCaffe focuses on taking advantage of sparsity. Also, you may want to compare the performance/portability with Winograd implementation in libxsmm, which is already used in TensorFlow.

**Comment by hfp** [comment]
Regarding Joongso's comment on Winograd in LIBXSMM, I recommend to build the library and the sample code under `samples/dnn` and then execute for example:

```
WINOGRAD=1 ./run_googlenetv1.sh
```
This will run all GoogleNetv1 convolutions forcing the Winograd algorithm (even when not beneficial). The default as used by the DNN sample codes relies on direct convolutions. However, a real use case would be to rely on LIBXSMM_DNN_CONV_ALGO_AUTO rather then LIBXSMM_DNN_CONV_ALGO_DIRECT or LIBXSMM_DNN_CONV_ALGO_WINOGRAD. The environment variable `WINOGRAD` is simply to reuse the sample code programs for testing purpose.

All sample codes print performance measures, which can be used to compare with your implementation. Please note that you may want to use at least an AVX2-capable system since otherwise only fallback code paths are taken. Further, since we merged some code which partially broke our Winograd implementation (whoever ;-) you may want to take LIBXSMM/20d993e:

```
git clone https://github.com/hfp/libxsmm.git
git checkout 20d993e
```

**Comment by imniksha** [comment]
Hello, I am new to Caffe and deep learning in general, and am hoping to find some answers here :)

I have installed SkimCaffe on my Ubuntu VM and am able to run classification models using Lenet. Now, I want to switch the convolution layer to winograd convolution layer, and perform comparative study between the two types of convolutions. 

I have tried to add it as below, however this has not been successful. The winograd layer addition just zeros all entries in the matrix and gives wrong classifications (refer below). I believe I must be doing something wrong here. I would greatly appreciate if someone could guide me to the solution.

Basically, I want to add a Winograd Layer (winograd convolution) into Lenet, using winograd_layer.cpp.

Thank you for the help! Also, please let me know where I could ask this question, if this is not the right platform for it :)

LENET:
layer {
  name: "train-data"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  transform_param {
    mean_file: "/home/x/DIGITS/digits/jobs/20180302-235120-dbc4/mean.binaryproto"
  }
  data_param {
    source: "/home/x/DIGITS/digits/jobs/20180302-235120-dbc4/train_db"
    batch_size: 64
    backend: LMDB
  }
}
layer {
  name: "val-data"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  transform_param {
    mean_file: "/home/x/DIGITS/digits/jobs/20180302-235120-dbc4/mean.binaryproto"
  }
  data_param {
    source: "/home/x/DIGITS/digits/jobs/20180302-235120-dbc4/val_db"
    batch_size: 32
    backend: LMDB
  }
}
layer {
  name: "scale"
  type: "Power"
  bottom: "data"
  top: "scaled"
  power_param {
    scale: 0.0125000001863
  }
}
layer {
  name: "win1"
  type: "Winograd"
  bottom: "scaled"
  top: "win1"
  param {
    lr_mult: 1.0
  }
  param {
    lr_mult: 2.0
  }
  convolution_param {
    num_output: 20
    kernel_size: 5
    stride: 1
  }
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "win1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "win2"
  type: "Winograd"
  bottom: "pool1"
  top: "win2"
  param {
    lr_mult: 1.0
  }
  param {
    lr_mult: 2.0
  }
  convolution_param {
    num_output: 50
    kernel_size: 5
    stride: 1
  }
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "win2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "ip1"
  type: "InnerProduct"
  bottom: "pool2"
  top: "ip1"
  param {
    lr_mult: 1.0
  }
  param {
    lr_mult: 2.0
  }
  inner_product_param {
    num_output: 500
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "ip1"
  top: "ip1"
}
layer {
  name: "ip2"
  type: "InnerProduct"
  bottom: "ip1"
  top: "ip2"
  param {
    lr_mult: 1.0
  }
  param {
    lr_mult: 2.0
  }
  inner_product_param {
    num_output: 10
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "ip2"
  bottom: "label"
  top: "accuracy"
  include {
    phase: TEST
  }
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "ip2"
  bottom: "label"
  top: "loss"
}

CLASSIFICATION OUTPUT:
image

**Comment by imniksha** [comment]
Hello, 

Have you tried substituting a Winograd Layer in for the Convolution Layer in a CNN using SkimCaffe yet?

I am doing a comparative study between the two types of convolution and want to insert a Winograd Conv Layer instead of the conv layer in Lenet. I have tried with unsuccessful results. 

Is this possible to do using SkimCaffe's winograd Layer? Could you please guide me to an example usage of the Winograd layer? I appreciate the help!

Thank you :)
